### PR TITLE
Disable no floating promises rule & explicit module boundary rule

### LINKED ISSUE #22: Update ESlint Config Rules
#### Describe the desired behavior
Update default eslint rules to be compatible with new versions of @typescript-eslint/parser & @typescript-eslint/eslint-plugin, both versions ^3.0.0.

#### Describe the current behavior
Our rules target the default rules found in version ^2.0.0.
As a result, our pxb-eslint-config is now recommending rules that we no longer have vetted as recommended or use within our own component libraries. 

#### Is this request related to a current issue?
No

### PR DESCRIPTION
**Pull request by emclaug2**
Fixes #22 

#### Changes proposed in this Pull Request
- Modifies @typescript-eslint/explicit-module-boundary-types to report an error instead of a warning.
- Disables @typescript-eslint/no-floating-promises rule.

It seems like the leading underscore syntax when injecting services in angular is getting more popular.  Syntax below:

`constructor(private _router: Router) ... `

The new version of @typescript-eslint is disregarding our @typescript-eslint/naming-convention parameter overrides and is requiring a leading underscore when we inject a service.  I think it might be a good time for us to adopt the leading underscore syntax when we inject our services.

### CODE REVIEW

**Comment by emclaug2** [comment]
These two rules are the ones I've observed after attempting to lint the dev branch of the angular-design patterns.  I'm going to attempt to lint the ng component library and see what happens there; maybe there are more rules to adjust.

**Comment by emclaug2** [comment]
Angular component library works just fine with these new rules, just needs minor tweaks.

**Comment by joebochill** [comment]
I kind of think that we should have the no-floating-promises rule enabled, no? What are we doing in our code that is throwing a warning about that?

And what exactly (in layman's terms) is the explicit module boundary types?

I agree with adopting the underscore syntax for private variables/injected services.

> The new version of @typescript-eslint is disregarding our @typescript-eslint/naming-convention parameter overrides
Can you elaborate on this? These rules were newly added, so I'm curious why they're being ignored.

**Comment by emclaug2** [comment]
> I kind of think that we should have the no-floating-promises rule enabled, no? What are we doing in our code that is throwing a warning about that?
> 
> And what exactly (in layman's terms) is the explicit module boundary types?
> 
> I agree with adopting the underscore syntax for private variables/injected services.
> 
> > The new version of @typescript-eslint is disregarding our @typescript-eslint/naming-convention parameter overrides
> 
> Can you elaborate on this? These rules were newly added, so I'm curious why they're being ignored.

The no-floating-promises-rule can just says that any promise must have a then/catch block afterwards, or be labeled with a `void`.  This rule mostly just breaks our unit tests in angular since they use a lot of promises.  

Explicit-module-boundary-types is the same thing as no-implicit-types.  Each arg in your function, and your functions signature must have types defined.

I don't know why the naming-convention is being ignored, I looked into the syntax for overriding the default behavior and it looks like we're inline with what's expected.  This might be a bug in their latest release perhaps?

**Comment by emclaug2** [comment]
> > I kind of think that we should have the no-floating-promises rule enabled, no? What are we doing in our code that is throwing a warning about that?
> > And what exactly (in layman's terms) is the explicit module boundary types?
> > I agree with adopting the underscore syntax for private variables/injected services.
> > > The new version of @typescript-eslint is disregarding our @typescript-eslint/naming-convention parameter overrides
> > 
> > 
> > Can you elaborate on this? These rules were newly added, so I'm curious why they're being ignored.
> 
> The no-floating-promises-rule says that any promise must have a then/catch block afterwards, otherwise it must be labeled with a `void`. This rule mostly just breaks our unit tests in angular since they use a lot of promises.  Every `expect` will need to be prefaced with a void and within the context of unit tests doesn't really add any value in my eyes.  I'm cool with enabling it if this is something you'd like to see added because it can be useful in other places.
> 
> Explicit-module-boundary-types is the same thing as no-implicit-types. Each arg in your function, and your functions signature must have types defined.
> 
> I don't know why the naming-convention is being ignored, I looked into the syntax for overriding the default behavior and it looks like we're inline with what's expected. This might be a bug in their latest release perhaps?

**Comment by joebochill** [comment]
> The no-floating-promises-rule says that any promise must have a then/catch block afterwards, otherwise it must be labeled with a void. This rule mostly just breaks our unit tests in angular since they use a lot of promises. Every expect will need to be prefaced with a void and within the context of unit tests doesn't really add any value in my eyes. I'm cool with enabling it if this is something you'd like to see added because it can be useful in other places.

In this case, I'd be inclined to keep the rule enabled (we don't want floating promises in our application/component code) and just have a `// eslint-disable no-floating-promises` comment at the top of our test/spec files where we don't actually need to handle all of these cases.

**Comment by emclaug2** [comment]
Ok I've updated the rule to enable ('error') the no-floating-promises rule.

**Comment by emclaug2** [comment]
This isn't ready to merge yet, just found a conflicting rule.   Our services are expected to have leading underscores, but if they are more than one word long (e.g. _drawerService) we get hit with this error: 

`error  Parameter Property name `_drawerService` must match one of the following formats: camelCase  @typescript-eslint/naming-convention`

I'll look into why the naming convention is not working now.

**Comment by emclaug2** [comment]
I loosened the naming-convention rules to allow for underscores on camelCase variables.

This is the problem I was facing: 
`constructor(public readonly _drawerService)`

 For some reason (still unknown to me) an injected `public` parameter was not allowed to use an underscore, but if you switched the modifier to `private` it'll work just fine.

**Comment by emclaug2** [comment]
> I loosened the naming-convention rules to allow for underscores on camelCase variables.
> 
> This is the problem I was facing:
> `constructor(public readonly _drawerService)`
> 
> For some reason (still unknown to me) an injected `public` parameter was not allowed to use an underscore, but if you switched the modifier to `private` it'll work just fine.

I realized this change isn't necessary.  What I should have done is just made the service private in the first place.  Good to merge I think.
